### PR TITLE
ISSUE-147 PGP: ClearEmailContent => EncryptedEmailContent

### PR DESCRIPTION
**Pull request by vttranlina**
Ref: https://github.com/linagora/tmail-backend/issues/174

### CODE REVIEW

File: tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/EncryptedEmailContent.scala [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+      name = parsedAttachment.getName.toScala,
+      contentType = parsedAttachment.getContentType.asString(),
+      cid = parsedAttachment.getCid
+        .toScala
+        .map(cid => cid.getValue),
+      isLine = parsedAttachment.isInline,
+      size = parsedAttachment.getContent.size())
+
+  def fromJava(listParsedAttachment: List[ParsedAttachment], messageId: MessageId): List[AttachmentMetadata] = {
+    val position: AtomicInteger = new AtomicInteger(0)
```
Comment (vttranlina) [inline tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/EncryptedEmailContent.scala:70]: I don't feel sure about this way (how to get `position` property)

File: tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/EncryptedEmailContent.scala [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+      encryptedHtml = encrypt(clearEmailContent.html),
+      hasAttachment = clearEmailContent.hasAttachment,
+      encryptedAttachmentMetadata = encryptAttachmentMetadata(clearEmailContent.attachments, messageId),
+      encryptedAttachmentContents = encryptAttachmentContent(clearEmailContent.attachments))
+
+  def encryptPreview(preview: Preview): String =
+    encrypt(preview.getValue)
+
+  def encryptAttachmentMetadata(parsedAttachments: List[ParsedAttachment], messageId: MessageId): Option[String] = {
+    if (parsedAttachments.isEmpty) return None
```
Comment (Arsnael) [inline tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/EncryptedEmailContent.scala:26]: You should use the pattern case matching syntax here :)

File: tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/EncryptedEmailContent.scala [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    Some(encrypt(AttachmentMetaDataSerializer
+      .serializeList(AttachmentMetadata.fromJava(parsedAttachments, messageId))
+      .toString()))
+  }
+
+  def encryptAttachmentContent(parsedAttachments: List[ParsedAttachment]): List[String] = {
+    if (parsedAttachments.isEmpty) return List.empty
```
Comment (Arsnael) [inline tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/EncryptedEmailContent.scala:33]: idem

File: tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/EncryptedEmailContent.scala [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+  def encrypt(byteSource: ByteSource): String = {
+    val stream: ByteArrayOutputStream = new ByteArrayOutputStream
+    encrypter.encrypt(byteSource, stream)
+    new String(stream.toByteArray, StandardCharsets.UTF_8)
+  }
+
+  def encrypt(value: String): String = {
+    val stream: ByteArrayOutputStream = new ByteArrayOutputStream
+    encrypter.encrypt(ByteSource.wrap(value.getBytes(StandardCharsets.UTF_8)), stream)
```
Comment (Arsnael) [inline tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/EncryptedEmailContent.scala:45]: you can just call `encrypt(byteSource: ByteSource)` with your string converted into a ByteSource

File: tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/EncryptedEmailContent.scala (hunk 1 above)
Comment (Arsnael) [reply tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/EncryptedEmailContent.scala:70]: I think position is really what it says in the ticket. Like the first attachment will have position 0, second position 1, etc

Which logically should allow you to match with the correct encrypted attachment content.

Like the attachment with position 0 in its metadata, you access its content with `encryptedAttachmentContents[0]`, etc

File: tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/EncryptedEmailContent.scala [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+      encryptedAttachmentMetadata = encryptAttachmentMetadata(clearEmailContent.attachments, messageId),
+      encryptedAttachmentContents = encryptAttachmentContent(clearEmailContent.attachments))
+
+  def encryptPreview(preview: Preview): String =
+    encrypt(preview.getValue)
+
+  def encryptAttachmentMetadata(parsedAttachments: List[ParsedAttachment], messageId: MessageId): Option[String] = {
+    if (parsedAttachments.isEmpty) return None
+    Some(encrypt(AttachmentMetaDataSerializer
+      .serializeList(AttachmentMetadata.fromJava(parsedAttachments, messageId))
```
Comment (Arsnael) [inline tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/EncryptedEmailContent.scala:28]: To what I said previously I would then create at the beginning of this method an atomic integer, that you start at -1, then you pass each time here the integer and you increment it same time (like using `incrementAndGet()` directly as an argument).

Then it allows you to remove your method `def fromJava(listParsedAttachment: List[ParsedAttachment], messageId: MessageId)` as you would not need it anymore

File: tmail-backend/mailbox/encrypted/api/src/test/scala/com/linagora/tmail/encrypted/AttachmentMetaDataSerializerTest.scala [hunk 6]
```diff
@@ -0,0 +1,78 @@
+package com.linagora.tmail.encrypted
+
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.Test
+import play.api.libs.json.{JsValue, Json}
+
+class AttachmentMetaDataSerializerTest {
+
+  @Test
+  def serializeShouldBeSuccess(): Unit = {
```
Comment (Arsnael) [inline tmail-backend/mailbox/encrypted/api/src/test/scala/com/linagora/tmail/encrypted/AttachmentMetaDataSerializerTest.scala:10]: You have a problem with those `Be` everywhere :) Remove it please ^^ (and in other tests names too)

File: tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/EncryptedEmailContent.scala (hunk 1 above)
Comment (vttranlina) [reply tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/EncryptedEmailContent.scala:70]: So, totally, we need to add one more property into `ParsedAttachment` (position), and need to refactor a `MessageParser`. Am I right?

Comment (Arsnael) [reply tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/EncryptedEmailContent.scala:70]: Nope, check this comment: https://github.com/linagora/tmail-backend/pull/175#discussion_r657613244 :)

File: tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/EncryptedEmailContent.scala (hunk 2 above)
Comment (vttranlina) [reply tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/EncryptedEmailContent.scala:26]: Sometimes, I feel confused between the two options. When I tried to use a matching pattern or monad, somebody told me I should use if-else to understand easily. And When I tried to if-else, somebody said to me should use functional style. :(

Comment (Arsnael) [reply tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/EncryptedEmailContent.scala:26]: Ahah yeah but on this case I think a matching pattern is better sorry^^'

File: tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/EncryptedEmailContent.scala (hunk 4 above)
Comment (vttranlina) [reply tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/EncryptedEmailContent.scala:45]: Sorry, I don't understand

Comment (Arsnael) [reply tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/EncryptedEmailContent.scala:53]:
```
def encrypt(value: String): String =
  encrypt(ByteSource.wrap(value.getBytes(StandardCharsets.UTF_8)))
```

File: tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/EncryptedEmailContent.scala [hunk 7]
```diff
@@ -23,16 +24,22 @@ class EncryptedEmailContentFactory(encrypter: Encrypter) {
     encrypt(preview.getValue)
 
   def encryptAttachmentMetadata(parsedAttachments: List[ParsedAttachment], messageId: MessageId): Option[String] = {
-    if (parsedAttachments.isEmpty) return None
-    Some(encrypt(AttachmentMetaDataSerializer
-      .serializeList(AttachmentMetadata.fromJava(parsedAttachments, messageId))
-      .toString()))
+    parsedAttachments match {
+      case Nil | List() => None
```
Comment (Arsnael) [inline tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/EncryptedEmailContent.scala:28]: Nil is not enough?

Comment (vttranlina) [reply tmail-backend/mailbox/encrypted/api/src/main/scala/com/linagora/tmail/encrypted/EncryptedEmailContent.scala:28]: Right, Not enough, If `parsedAttachments` is List(), We can still json deserialize that, so `encryptAttachmentMetadata` will has value

**Review by Arsnael — APPROVED** [review]

**Pull request merged**